### PR TITLE
Kernel: Basic AArch64 exception handling #10468

### PR DESCRIPTION
**Pull request by Quaker762**
This provides a basic mechanism to install a Vector Table for Exception Level 1, as well as a basic "kernel entry" assembly routine in which the current processor context is saved to the stack and passed into the common exception C++ function.

Remake of #10468

### CODE REVIEW

**Comment by bgianfo** [comment]
CC: @jamesmintram, @nico, and @ADKaster who had looked at the previous PR.

**Comment by Quaker762** [comment]
I've fucked up the rebase on this one big time. Moving to draft for the minute.

File: Kernel/Prekernel/Arch/aarch64/vector_table.S [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    // Exceptions from Lower EL, where causing application is in AArch32 mode
+    unimplemented_entry
+    unimplemented_entry
+    unimplemented_entry
+    unimplemented_entry
+
+synchronous_current_elsp_elx:
+    save_current_context
+    restore_previous_context
+    eret
```
Comment (FireFox317) [inline Kernel/Prekernel/Arch/aarch64/vector_table.S:132]: By looking at this code, it is not clear that `save_current_context` also jumps to the `exception_common` function. I would suggest to move the `bl exception_common` out of `save_current_context`, and put it in each function instead. Or to reword the `save_current_context` function, such that is is clear that it also jumps somewhere.

Comment (Quaker762) [reply Kernel/Prekernel/Arch/aarch64/vector_table.S:132]: Good point, I'll move it out of the macro and call it directly from the label

File: Kernel/Prekernel/Arch/aarch64/vector_table.S [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    // Move stack pointer into first argument register
+    // and jump to the C++ exception handler
+    mov x0, sp
+    bl exception_common
+.endm
+
+.macro restore_previous_context
+    // Restore special registers first
+    ldr x0, [sp, #SPSR_EL1_SLOT]
```
Comment (ADKaster) [inline Kernel/Prekernel/Arch/aarch64/vector_table.S:73]: This doesn't look like it restores tpidr_el0 or sp_el0 from the trap frame.

Is that intentional? If so, could we add a comment saying like "we saved tpidr_el0 and sp_el0 for debugging purposes, FIXME shrink the stack frame to only required variables"?

Comment (Quaker762) [reply Kernel/Prekernel/Arch/aarch64/vector_table.S:73]: This is a bug. I've restored this register register too

**Pull request closed without merging**